### ISSUE TITLE
Refactor Translation storage component

### ISSUE THREAD

**Issue by eternoendless** [PrestaShop/PrestaShop#19242]
## Quickmap

See below for details

- [x] https://github.com/PrestaShop/PrestaShop/issues/23761 Introduce structure
- [x] https://github.com/PrestaShop/PrestaShop/issues/23762 Make a single provider for Core translations which have the same logic. Add all other Providers. Introduce ProviderFactory.
- [x] https://github.com/PrestaShop/PrestaShop/issues/23764 Restore initial component
- [x] https://github.com/PrestaShop/PrestaShop/pull/23621 Restructure `src/Core/Translation/` 
- [x] https://github.com/PrestaShop/PrestaShop/issues/23763 Plug Translation BO Interface on new Storage component
- [ ] https://github.com/PrestaShop/PrestaShop/issues/23765 Caching system for ThemeExtractor
- [x] https://github.com/PrestaShop/PrestaShop/issues/23766 Fix Theme export - QA needed 
- [x] #23344 Error when adding new theme

#### Is your feature request related to a problem?

The implementation of the translation feature in BO* is excessively complex and has [several problems](https://github.com/PrestaShop/PrestaShop/issues/12873), including unclear interfaces and responsibilities, not to mention bugs (like [this one](https://github.com/PrestaShop/PrestaShop/issues/16555)). 

#### Describe the solution you'd like

Functional specification: https://github.com/PrestaShop/prestashop-specs/pull/120

The backend should be refactored to allow the feature to work as specified, with a leaner, saner structure.

## Deeper in the refactoring (edited by @matks )

We now think the Translation feature can be splitted into 3 components
- the **Translator** component: in charge of translating a string for a given locale (`$translator->translate(...)`)
- the **Translation Storage** component: in charge of building the Catalogues needed by the Translator from wherever they are (there are many sources, depending if it's for the Back office, the Front office, the modules...)
- the **Admin Interface**: Back Office pages which provide the ability to see messages that can be translated and save user-defined messages

This refactoring only targets the Storage component. This refactoring aims to build a new, better implementation of the Storage component then unplug the Admin Interface from the old component to the new component.

The old component must stay unmodified until step

### New Storage component

We are going to build a new Storage component in `src/Core/Translation/{...}`. (`{...}` = Storage?)
It is mostly built around Providers, one per usecase. A Provider must provide for a given context (= module, theme, FO, BO...) the 3 catalogues (default, translated, user-defined).

### Roadmap to this result

1. https://github.com/PrestaShop/PrestaShop/issues/23761 : Introduce structure for new component with one Provider and one factor - Unit tested - No QA - MERGED ✅ 

_Note: the above PR mistakenly introduced elements from the Admin Interface. It also modified some files from the old component, this will be fixed by step 4._

2. Make a single provider for Core translations which have the same logic. Add all other Providers - Unit tested - No QA - https://github.com/PrestaShop/PrestaShop/issues/23762 MERGED ✅ 
3. Introduce ProviderFactory to target a single provider by translation **type** - Unit tested - No QA - https://github.com/PrestaShop/PrestaShop/issues/23762 MERGED ✅ 

4. Restore initial component like it was in PrestaShopBundle - No QA https://github.com/PrestaShop/PrestaShop/issues/23764

5. Restructure `src/Core/Translation/` to comply with new structure - https://github.com/PrestaShop/PrestaShop/pull/23621 MERGED ✅ 

6. Plug Translation BO Interface on new Storage component - QA needed 😬  https://github.com/PrestaShop/PrestaShop/issues/23763

7. Caching system for ThemeExtractor - Unit tested - No QA https://github.com/PrestaShop/PrestaShop/issues/23765

8. Fix Theme export - QA needed https://github.com/PrestaShop/PrestaShop/issues/23766

**Related bugs**

- [x] https://github.com/PrestaShop/PrestaShop/issues/23344
- [x] https://github.com/PrestaShop/PrestaShop/issues/23901

**Nice-to-have (not blocking for feature freeze)**

9. Map old component usage - build a documentation of how Translation is used https://github.com/PrestaShop/PrestaShop/issues/23767
8. Wipe out old component or deprecate it - QA needed https://github.com/PrestaShop/PrestaShop/issues/23767
10. Write doc about the new component https://github.com/PrestaShop/PrestaShop/issues/23767

**Comment by eternoendless** [comment]
Unfortunately this refactor is taking too long and will have to be bumped for 1.7.8 in order to unblock 1.7.7

**Comment by sowbiba** [comment]
Since the review of the big PR is difficult because so many things have been modified, the PR is divided by multiple dependant parts to ease the review

The parts will be :

**1.7.8 MVP**
- [x] Replace Translations providers pass by single service and get rid of AbstractProvider - Unit tested - No QA https://github.com/PrestaShop/PrestaShop/issues/23761
- [x] Make a single provider for Core translations which have the same logic - Unit tested - No QA https://github.com/PrestaShop/PrestaShop/issues/23762
- [x] Introduce ProviderFactory to target a single provider by translation **type** - Unit tested - No QA https://github.com/PrestaShop/PrestaShop/issues/23762
- [x] Restore initial component like it was in PrestaShopBundle - No QA https://github.com/PrestaShop/PrestaShop/issues/23764
- [x] Improve folder structure - Unit tested - No QA  https://github.com/PrestaShop/PrestaShop/pull/23621
- [x] Plug Translation BO Interface on new Storage component - QA needed https://github.com/PrestaShop/PrestaShop/issues/23763
- [x] Fix Theme export - QA needed https://github.com/PrestaShop/PrestaShop/issues/23766
- [ ] Caching system for ThemeExtractor - Unit tested - No QA https://github.com/PrestaShop/PrestaShop/issues/23765

**1.7.9 if needed**

- [ ] Map old component usage - build a documentation of how Translation is used https://github.com/PrestaShop/PrestaShop/issues/23767
- [ ] Wipe out old component or deprecate it - QA needed https://github.com/PrestaShop/PrestaShop/issues/23767
- [ ] Write doc about the new component https://github.com/PrestaShop/PrestaShop/issues/23767

Roadmap remapped to top comment https://github.com/PrestaShop/PrestaShop/issues/19242

**Comment by matks** [comment]
Reopening as https://github.com/PrestaShop/PrestaShop/issues/23765 is left for 1.7.9
